src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilerCache.cs

```diff
@@ -27,178 +25,78 @@ public class CompilerCache : ICompilerCache
                 {
-                    return null;
-                }
+                    cacheResultToCache = CompilerCacheResult.FileNotFound;
+                    cacheResult = CompilerCacheResult.FileNotFound;
 
-                var relativeFileInfo = new RelativeFileInfo(fileInfo, relativePath);
-                return OnCacheMiss(relativeFileInfo, normalizedPath, compile);
-            }
-            else if (cacheEntry.IsPreCompiled && !cacheEntry.IsValidatedPreCompiled)
-            {
-                // For precompiled views, the first time the entry is read, we need to ensure that no changes were made
-                // either to the file associated with this entry, or any _ViewImports associated with it between the time
-                // the View was precompiled and the time EnsureInitialized was called. For later iterations, we can
-                // rely on expiration triggers ensuring the validity of the entry.
-
-                var fileInfo = _fileProvider.GetFileInfo(relativePath);
-                if (!fileInfo.Exists)
-                {
-                    return null;
-                }
-
-                var relativeFileInfo = new RelativeFileInfo(fileInfo, relativePath);
-                if (cacheEntry.Length != fileInfo.Length)
-                {
-                    // Recompile if the file lengths differ
-                    return OnCacheMiss(relativeFileInfo, normalizedPath, compile);
+                    cacheEntryOptions = new MemoryCacheEntryOptions();
+                    cacheEntryOptions.AddExpirationTrigger(_fileProvider.Watch(relativePath));
                 }
-
-                if (AssociatedGlobalFilesChanged(cacheEntry, compile))
-                {
-                    // Recompile if _ViewImports have changed since the entry was created.
-                    return OnCacheMiss(relativeFileInfo, normalizedPath, compile);
-                }
-
-                if (cacheEntry.LastModified == fileInfo.LastModified)
+                else
                 {
-                    // Assigning to IsValidatedPreCompiled is an atomic operation and will result in a safe race
-                    // if it is being concurrently updated and read.
-                    cacheEntry.IsValidatedPreCompiled = true;
-                    return new GetOrAddResult
-                    {
-                        CompilationResult = CompilationResult.Successful(cacheEntry.CompiledType),
-                        CompilerCacheEntry = cacheEntry
-                    };
+                    var relativeFileInfo = new RelativeFileInfo(fileInfo, relativePath);
+                    var compilationResult = compile(relativeFileInfo).EnsureSuccessful();
+                    cacheEntryOptions = GetMemoryCacheEntryOptions(relativePath);
+
+                    // Don't cache compilationResult since it could be UncachedCompilationResult
+                    // which has the generated view content as a string.
```

Comment:
I feel like this needs to be expanded a bit. Under what conditions might this happen?